lib/cfg-tree.c

```diff
@@ -1333,6 +1333,50 @@ cfg_tree_check_inline_template(CfgTree *self, const gchar *template_or_name, GEr
   return template;
 }
 
+gboolean
+_count_sources(LogExprNode *node, gpointer user_data)
+{
+  gint *counter = (gint *)user_data;
+
+  if (node->content == ENC_SOURCE)
+    ++(*counter);
+
+  return TRUE;
+}
+
+static void
+cfg_tree_foreach(LogExprNode *node, gboolean (*func)(LogExprNode *, gpointer user_data), gpointer user_data)
+{
+  g_assert(func != NULL);
+  g_assert(node != NULL);
+
+  if (!func(node, user_data))
+    return;
+
+  if (node->children)
+    cfg_tree_foreach(node->children, func, user_data);
+
+  if (node->next)
+    cfg_tree_foreach(node->next, func, user_data);
+}
+
+static gboolean
+cfg_tree_check_empty_pipelines(CfgTree *self)
+{
+  gint source_counter = 0;
+
+  for (gint i = 0; i < self->rules->len; i++)
```

Comment:
And the for loop itself is cfg_tree_foreach(), while the remaining logic is the check_empty_pipelines() function.